prov/verbs/src/verbs_mr.c

```diff
@@ -88,79 +164,138 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
 		fi_ibv_access |= IBV_ACCESS_LOCAL_WRITE;
 
 	/* Local read access to an MR is enabled by default in verbs */
-	if (access & FI_RECV)
+	if (attr->access & FI_RECV)
 		fi_ibv_access |= IBV_ACCESS_LOCAL_WRITE;
 
 	/* iWARP spec requires Remote Write access for an MR that is used
 	 * as a data sink for a Remote Read */
-	if (access & FI_READ) {
+	if (attr->access & FI_READ) {
 		fi_ibv_access |= IBV_ACCESS_LOCAL_WRITE;
 		if (md->domain->verbs->device->transport_type == IBV_TRANSPORT_IWARP)
 			fi_ibv_access |= IBV_ACCESS_REMOTE_WRITE;
 	}
 
-	if (access & FI_WRITE)
+	if (attr->access & FI_WRITE)
 		fi_ibv_access |= IBV_ACCESS_LOCAL_WRITE;
 
-	if (access & FI_REMOTE_READ)
+	if (attr->access & FI_REMOTE_READ)
 		fi_ibv_access |= IBV_ACCESS_REMOTE_READ;
 
 	/* Verbs requires Local Write access too for Remote Write access */
-	if (access & FI_REMOTE_WRITE)
+	if (attr->access & FI_REMOTE_WRITE)
 		fi_ibv_access |= IBV_ACCESS_LOCAL_WRITE |
-			IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_ATOMIC;
+				 IBV_ACCESS_REMOTE_WRITE |
+				 IBV_ACCESS_REMOTE_ATOMIC;
 
-	md->mr = ibv_reg_mr(md->domain->pd, (void *) buf, len, fi_ibv_access);
-	if (!md->mr)
+	ret = fi_ibv_mr_reg_common(md, fi_ibv_access, attr->mr_iov[0].iov_base,
+				   attr->mr_iov[0].iov_len, attr->context);
+	if (OFI_UNLIKELY(ret))
 		goto err;
 
-	md->mr_fid.mem_desc = (void *) (uintptr_t) md->mr->lkey;
-	md->mr_fid.key = md->mr->rkey;
 	*mr = &md->mr_fid;
-	if (md->domain->eq_flags & FI_REG_MR) {
-		struct fi_eq_entry entry = {
-			.fid = &md->mr_fid.fid,
-			.context = context
-		};
-		if (md->domain->eq)
-			fi_ibv_eq_write_event(md->domain->eq, FI_MR_COMPLETE,
-				 	      &entry, sizeof(entry));
-		else if (md->domain->util_domain.eq)
-			 /* This branch is taken for the verbs/DGRAM */
-			fi_eq_write(&md->domain->util_domain.eq->eq_fid,
-				    FI_MR_COMPLETE, &entry, sizeof(entry), 0);
-	}
-	return 0;
-
+	return FI_SUCCESS;
 err:
 	free(md);
-	return -errno;
+	return -ret;
```

Comment:
This should just be ret. fi_ibv_mr_reg_common already negates errno when it returns.

Reply:
Good catch, thanks!
Done